Fix render tree construction when inlines are inserted in reverse after a block in a continuation

Ahmad-S792 opened the pull request:
#### 4af8f052cd4fee2c1f169220ab61ed94f512d0ff
<pre>
Fix render tree construction when inlines are inserted in reverse after a block in a continuation

<a href="https://bugs.webkit.org/show_bug.cgi?id=274762">https://bugs.webkit.org/show_bug.cgi?id=274762</a>
<a href="https://rdar.apple.com/problem/128826228">rdar://problem/128826228</a>

Reviewed by Darin Adler.

This patch aligns WebKit with Blink / Chromium and
Gecko / Firefox.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/6ac676e5e68e3271321b9c8236192b6a530d055e">https://chromium.googlesource.com/chromium/blink/+/6ac676e5e68e3271321b9c8236192b6a530d055e</a>

When inserting inlines after a block inside of an inline continuation,
if we insert inline content in reverse DOM order, we can (depending on
when we attach) end up appending the renderers in the wrong location,
leading to visual aberrations.

The fix is in RenderTreeBuilder, where we should always attach behind our
beforeChild if we&apos;re inserting an inline and beforeChild is an inline,
regardless of other continuation state.

* Source/WebCore/rendering/updating/RenderTreeBuilderInline.cpp:
(WebCore::RenderTreeBuilder::Inline::insertChildToContinuation):
* LayoutTests/fast/inline/continuation-inlines-inserted-in-reverse-after-block.html: Add Test Case
* LayoutTests/fast/inline/continuation-inlines-inserted-in-reverse-after-block-expected.txt: Add Test Case Expectation
* LayoutTests/platform/mac/tables/mozilla/bugs/bug113235-3-expected.txt: Rebaselined
* LayoutTests/platform/mac/tables/mozilla/bugs/bug113235-3-expected.png: Ditto
* LayoutTests/platform/glib/tables/mozilla/bugs/bug113235-3-expected.txt: Ditto
* LayoutTests/platform/ios/tables/mozilla/bugs/bug113235-3-expected.txt: Ditto

Canonical link: <a href="https://commits.webkit.org/279422@main">https://commits.webkit.org/279422@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70f7ffadd05d4292f6a78fbb2d5849a4d75ed437

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5973 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56748 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3973 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2744 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55566 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46198 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/24472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3518 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2350 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58343 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28625 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3690 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29832 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46394 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11647 "Found 1 new test failure: fast/dom/intersection-observer-document-leak.html (failure)") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7865 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29602 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/updating/RenderTreeBuilderInline.cpp:
```diff
@@ -1,5 +1,6 @@
 /*
- * Copyright (C) 2018 Apple Inc. All rights reserved.
+ * Copyright (C) 2018-2024 Apple Inc. All rights reserved.
+ * Copyright (C) 2013 Google Inc. All rights reserved.
```
I understand the desire to give credit to all contributors, but a single clause in an if statement is not really enough for a copyright update.

Ahmad-S792 replied:
Sure, will revert in next push and submit for 'merge-queue'.

Ahmad-S792 replied:
Done and pushed newer commit with following removed (and rebased). :-)

Commit:
Fix render tree construction when inlines are inserted in reverse after a block in a continuation

https://bugs.webkit.org/show_bug.cgi?id=274762
rdar://problem/128826228

Reviewed by Darin Adler.

This patch aligns WebKit with Blink / Chromium and
Gecko / Firefox.

Merge: https://chromium.googlesource.com/chromium/blink/+/6ac676e5e68e3271321b9c8236192b6a530d055e

When inserting inlines after a block inside of an inline continuation,
if we insert inline content in reverse DOM order, we can (depending on
when we attach) end up appending the renderers in the wrong location,
leading to visual aberrations.

The fix is in RenderTreeBuilder, where we should always attach behind our
beforeChild if we're inserting an inline and beforeChild is an inline,
regardless of other continuation state.

* Source/WebCore/rendering/updating/RenderTreeBuilderInline.cpp:
(WebCore::RenderTreeBuilder::Inline::insertChildToContinuation):
* LayoutTests/fast/inline/continuation-inlines-inserted-in-reverse-after-block.html: Add Test Case
* LayoutTests/fast/inline/continuation-inlines-inserted-in-reverse-after-block-expected.txt: Add Test Case Expectation
* LayoutTests/platform/mac/tables/mozilla/bugs/bug113235-3-expected.txt: Rebaselined
* LayoutTests/platform/mac/tables/mozilla/bugs/bug113235-3-expected.png: Ditto
* LayoutTests/platform/glib/tables/mozilla/bugs/bug113235-3-expected.txt: Ditto
* LayoutTests/platform/ios/tables/mozilla/bugs/bug113235-3-expected.txt: Ditto

Canonical link: https://commits.webkit.org/279422@main
